File: apps/demo-fs-ts-nextjs/test/unit/be-steps/token-management.steps.ts
import path from "path";
import { loadFeature, describeFeature } from "@amiceli/vitest-cucumber";
import { expect } from "vitest";
import { createTestContext, registerUser, loginUser, getAuth, type TestContext } from "./helpers/test-context";

const feature = await loadFeature(
  path.resolve(process.cwd(), "../../specs/apps/demo/be/gherkin/token-management/tokens.feature"),
);

async function setupAdmin(ctx: TestContext): Promise<void> {
  await registerUser(ctx, "superadmin", "superadmin@example.com", "Str0ng#Pass1");
  await ctx.client.dispatch("POST", "/api/v1/test/promote-admin", { username: "superadmin" }, null);
  await loginUser(ctx, "superadmin", "Str0ng#Pass1");
}

describeFeature(feature, ({ Scenario, Background }) => {
  let ctx: TestContext;
  let decodedPayload: Record<string, unknown>;

  Background(({ Given, And }) => {
    Given("the API is running", () => {
      ctx = createTestContext();
      decodedPayload = {};
    });

    And('a user "alice" is registered with password "Str0ng#Pass1"', async () => {
      await registerUser(ctx, "alice", "alice@example.com", "Str0ng#Pass1");
    });

    And('"alice" has logged in and stored the access token', async () => {
      await loginUser(ctx, "alice", "Str0ng#Pass1");
    });
  });

  Scenario("Access token payload contains user ID claim", ({ When, Then }) => {
    When("alice decodes her access token payload", () => {
      const token = ctx.tokens.get("alice_access")!;
      const parts = token.split(".");
      decodedPayload = JSON.parse(Buffer.from(parts[1]!, "base64url").toString());
    });

    Then('the token should contain a non-null "sub" claim', () => {
      expect(decodedPayload.sub).toBeDefined();
    });
  });

  Scenario("Access token payload contains issuer claim", ({ When, Then }) => {
    When("alice decodes her access token payload", () => {
      const token = ctx.tokens.get("alice_access")!;
      const parts = token.split(".");
      decodedPayload = JSON.parse(Buffer.from(parts[1]!, "base64url").toString());
    });

    Then('the token should contain a non-null "iss" claim', () => {
      expect(decodedPayload.iss).toBeDefined();
    });
  });

  Scenario("JWKS endpoint returns the public key for token signature verification", ({ When, Then, And }) => {
    When("the client sends GET /.well-known/jwks.json", async () => {
      ctx.response = await ctx.client.dispatch("GET", "/.well-known/jwks.json", null, null);
    });

    Then("the response status code should be 200", () => {
      expect(ctx.response!.status).toBe(200);
    });

    And('the response body should contain at least one key in the "keys" array', () => {
      const body = ctx.response!.body as { keys: unknown[] };
      expect(body.keys.length).toBeGreaterThanOrEqual(1);
    });
  });

  Scenario("Logout blacklists the access token", ({ When, Then, And }) => {
    When("alice sends POST /api/v1/auth/logout with her access token", async () => {
      ctx.response = await ctx.client.dispatch("POST", "/api/v1/auth/logout", null, getAuth(ctx, "alice"));
    });

    Then("the response status code should be 200", () => {
      expect(ctx.response!.status).toBe(200);
    });

    And("alice's access token should be recorded as revoked", async () => {
      const resp = await ctx.client.dispatch("GET", "/api/v1/users/me", null, getAuth(ctx, "alice"));
      expect(resp.status).toBe(401);
    });
  });

  Scenario("Blacklisted access token is rejected with 401 on protected endpoints", ({ Given, When, Then }) => {
    Given("alice has logged out and her access token is blacklisted", async () => {
      await ctx.client.dispatch("POST", "/api/v1/auth/logout", null, getAuth(ctx, "alice"));
    });

    When("the client sends GET /api/v1/users/me with alice's access token", async () => {
      ctx.response = await ctx.client.dispatch("GET", "/api/v1/users/me", null, getAuth(ctx, "alice"));
    });

    Then("the response status code should be 401", () => {
      expect(ctx.response!.status).toBe(401);
    });
  });

  Scenario("Deactivating a user revokes all their active tokens", ({ Given, And, When, Then }) => {
    Given('an admin user "superadmin" is registered and logged in', async () => {
      await setupAdmin(ctx);
    });

    And("the admin has disabled alice's account via POST /api/v1/admin/users/{alice_id}/disable", async () => {
      const aliceId = ctx.userIds.get("alice")!;
      await ctx.client.dispatch(
        "POST",
        `/api/v1/admin/users/${aliceId}/disable`,
        { reason: "Policy violation" },
        getAuth(ctx, "superadmin"),
      );
    });

    When("the client sends GET /api/v1/users/me with alice's access token", async () => {
      ctx.response = await ctx.client.dispatch("GET", "/api/v1/users/me", null, getAuth(ctx, "alice"));
    });

    Then("the response status code should be 401", () => {
      expect(ctx.response!.status).toBe(401);
    });
  });
});

File: apps/crud-be-python-fastapi/tests/unit/steps/token_management_steps.py
"""BDD step definitions for token management feature."""

import base64
import json

from pytest_bdd import given, parsers, scenarios, then, when

from tests.integration.service_client import FakeResponse, ServiceClient
from tests.integration.steps.security_steps import ADMIN_PASSWORD, register_and_promote_admin
from tests.unit.conftest import GHERKIN_ROOT

scenarios(str(GHERKIN_ROOT / "token-management" / "tokens.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    '"alice" has logged in and stored the access token',
    target_fixture="alice_tokens",
)
def alice_login_for_tokens(client: ServiceClient, registered_user: dict) -> dict:
    return client.login_user("alice", _PASSWORD)


# @covers specs/apps/crud/behavior/crud-be/gherkin/token-management/tokens.feature:Blacklisted access token is rejected with 401 on protected endpoints
@given("alice has logged out and her access token is blacklisted")
def alice_logout_blacklisted(client: ServiceClient, alice_tokens: dict) -> None:
    client.post_logout(f"Bearer {alice_tokens['accessToken']}")


@given(
    parsers.parse('an admin user "{username}" is registered and logged in'),
    target_fixture="admin_tokens",
)
def admin_login_for_tokens(client: ServiceClient, username: str) -> dict:
    user_data = register_and_promote_admin(client, username, ADMIN_PASSWORD)
    tokens = client.login_user(username, ADMIN_PASSWORD)
    return {**tokens, "id": user_data["id"]}


# @covers specs/apps/crud/behavior/crud-be/gherkin/token-management/tokens.feature:Deactivating a user revokes all their active tokens
@given("the admin has disabled alice's account via POST /api/v1/admin/users/{alice_id}/disable")
def admin_disable_alice(client: ServiceClient, registered_user: dict, admin_tokens: dict) -> None:
    resp = client.post_admin_disable_user(
        registered_user["id"],
        f"Bearer {admin_tokens['accessToken']}",
        reason="Test",
    )
    assert resp.status_code == 200, f"Disable failed: {resp.text}"


# --- When steps ---


@when("alice decodes her access token payload", target_fixture="decoded_payload")
def alice_decode_token(alice_tokens: dict) -> dict:
    token = alice_tokens["accessToken"]
    parts = token.split(".")
    assert len(parts) == 3, "Invalid JWT structure"
    payload_b64 = parts[1]
    padding = 4 - len(payload_b64) % 4
    if padding != 4:
        payload_b64 += "=" * padding
    return json.loads(base64.urlsafe_b64decode(payload_b64).decode("utf-8"))


@when("the client sends GET /.well-known/jwks.json", target_fixture="response")
def get_jwks(client: ServiceClient) -> FakeResponse:
    return client.get_jwks()


@when("alice sends POST /api/v1/auth/logout with her access token", target_fixture="response")
def alice_logout_tokens(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.post_logout(f"Bearer {alice_tokens['accessToken']}")


@when(
    "the client sends GET /api/v1/users/me with alice's access token",
    target_fixture="response",
)
def get_me_with_alice_token(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.get_me(f"Bearer {alice_tokens['accessToken']}")


# --- Then steps ---


# @covers specs/apps/crud/behavior/crud-be/gherkin/token-management/tokens.feature:Access token payload contains user ID claim
# @covers specs/apps/crud/behavior/crud-be/gherkin/token-management/tokens.feature:Access token payload contains issuer claim
@then(parsers.parse('the token should contain a non-null "{claim}" claim'))
def check_token_claim(decoded_payload: dict, claim: str) -> None:
    assert claim in decoded_payload, f"Claim '{claim}' not in token: {decoded_payload}"
    assert decoded_payload[claim] is not None, f"Claim '{claim}' is null"


# @covers specs/apps/crud/behavior/crud-be/gherkin/token-management/tokens.feature:JWKS endpoint returns the public key for token signature verification
@then('the response body should contain at least one key in the "keys" array')
def check_jwks_keys(response: FakeResponse) -> None:
    body = response.json()
    assert "keys" in body, f"No 'keys' field in JWKS response: {body}"
    assert len(body["keys"]) > 0, "JWKS keys array is empty"


# @covers specs/apps/crud/behavior/crud-be/gherkin/token-management/tokens.feature:Logout blacklists the access token
@then("alice's access token should be recorded as revoked")
def check_alice_token_revoked(client: ServiceClient, alice_tokens: dict) -> None:
    resp = client.get_me(f"Bearer {alice_tokens['accessToken']}")
    assert resp.status_code == 401, f"Expected 401 for revoked token, got {resp.status_code}"
